tensorrt_llm/quantization/layers.py

```diff
@@ -1069,8 +1082,51 @@ def forward(
                 dtype=dtype,
                 tp_size=self.tp_size,
                 tp_rank=self.tp_rank)
+        
+        def dequantize_tensor(x, scale, invert_scale = False):
+            # Cast from int8 to dtype
+            casted_x = cast(x, self.dtype)
+            if invert_scale:
+                return casted_x / scale
+            return casted_x * scale
+        
+        # if cross attention, cross QKV only needs to be calculated once in the
+        # 1st decoding step --> write to cross KV cache --> remains constant
+        # during the entire decoding. 1st and >1 steps are distinguished by
+        # whether past_key_value exists or not
+        # also, cross KV cache max length is set from encoder output seqlen,
+        # this maps to the max context length concept in decoder-only models
+        cross_qkv = None
+        # get length data in every run
+        if encoder_output:
+            assert isinstance(encoder_output, Tensor)
+        # but only do projection once at 1st decoding step
+        if self.cross_attention and encoder_output:
+            if self.quant_mode.has_act_and_weight_quant():
+                if self.quant_mode.has_act_static_scaling():
+                    # Avoid quantiztion layers as it breaks int8 plugins
+                    encoder_output = quantize_tensor(
```

Comment:
I am really uncertain if this is the right scaling factor for quantization, and even more uncertain about the factors for the dequantization step.

This also seems like it would be quite slow to do online, so I'm wondering if there is a better way of implementing this.